cmd/juju/commands/bootstrap.go

```diff
@@ -315,31 +315,290 @@ func (c *bootstrapCommand) Run(ctx *cmd.Context) (resultErr error) {
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	region, err := getRegion(cloud, c.Cloud, regionName)
+	if err != nil {
+		fmt.Fprintf(ctx.GetStderr(),
+			"%s\n\nSpecify an alternative region, or try %q.\n",
+			err, "juju update-clouds",
+		)
+		return cmd.ErrSilent
+	}
+	if c.controllerName == "" {
+		c.controllerName = defaultControllerName(c.Cloud, region.Name)
+	}
+
+	config, err := c.bootstrapConfigs(ctx, cloud, provider)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Read existing current controller so we can clean up on error.
+	var oldCurrentController string
+	store := c.ClientStore()
+	oldCurrentController, err = store.CurrentController()
+	if errors.IsNotFound(err) {
+		oldCurrentController = ""
+	} else if err != nil {
+		return errors.Annotate(err, "error reading current controller")
+	}
+
+	defer func() {
+		if resultErr == nil || errors.IsAlreadyExists(resultErr) {
+			return
+		}
+		if oldCurrentController != "" {
+			if err := store.SetCurrentController(oldCurrentController); err != nil {
+				logger.Errorf(
+					"cannot reset current controller to %q: %v",
+					oldCurrentController, err,
+				)
+			}
+		}
+		if err := store.RemoveController(c.controllerName); err != nil {
+			logger.Errorf(
+				"cannot destroy newly created controller %q details: %v",
+				c.controllerName, err,
+			)
+		}
+	}()
+
+	environ, err := bootstrapPrepare(
+		modelcmd.BootstrapContext(ctx), store,
+		bootstrap.PrepareParams{
+			ModelConfig:      config.bootstrapModel,
+			ControllerConfig: config.controller,
+			ControllerName:   c.controllerName,
+			Cloud: environs.CloudSpec{
+				Type:             cloud.Type,
+				Name:             c.Cloud,
+				Region:           region.Name,
+				Endpoint:         region.Endpoint,
+				IdentityEndpoint: region.IdentityEndpoint,
+				StorageEndpoint:  region.StorageEndpoint,
+				Credential:       credentials.credential,
+			},
+			CredentialName: credentials.name,
+			AdminSecret:    config.bootstrap.AdminSecret,
+		},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	hostedModelUUID, err := utils.NewUUID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Set the current model to the initial hosted model.
+	if err := store.UpdateModel(c.controllerName, c.hostedModelName, jujuclient.ModelDetails{
+		hostedModelUUID.String(),
+	}); err != nil {
+		return errors.Trace(err)
+	}
+	if err := store.SetCurrentModel(c.controllerName, c.hostedModelName); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Set the current controller so "juju status" can be run while
+	// bootstrapping is underway.
+	if err := store.SetCurrentController(c.controllerName); err != nil {
+		return errors.Trace(err)
+	}
+
+	cloudRegion := c.Cloud
+	if region.Name != "" {
+		cloudRegion = fmt.Sprintf("%s/%s", cloudRegion, region.Name)
+	}
+	ctx.Infof(
+		"Creating Juju controller %q on %s",
+		c.controllerName, cloudRegion,
+	)
+
+	// If we error out for any reason, clean up the environment.
+	defer func() {
```

Comment:
This deferred closure in particular is extremely difficult to understand (I remember running into this when I first started). We should definitely pursue inlining this since it should only be necessary after the environment has begun bootstrapping.

I believe the reason we've written it this way is to avoid having to replicate the call to `destroyEnvironment` (there are three error-cases which would have to be handled). I think this is a hint that this function is (still) doing too much. We should have a function that does what it can to prepare an environment `maybeBootstrapEnvironment` which returns the necessary information to tear an environment down, and an error. Then the error should be checked, and if the other variables are populated. Something along the lines of:

```go
environ, err := maybeBootstrapEnvironment(/*...*/)
if err != nil {
    if environ != nil && c.KeepBrokenEnvironment == false {
        environ.Teardown()
    }
}
```

This is so much clearer than a deferred closure with multiple conditionals.